src/pointer-analysis/value_set.cpp

```diff
@@ -569,12 +569,34 @@ void value_sett::get_value_set_rec(
     // Find out what the pointer operand points at, and suck that data into
     // new object maps.
     object_mapt op0_set;
-    if(!is_pointer_type(op1))
-      get_value_set_rec(op0, op0_set, "", op0->type, false);
-
     object_mapt op1_set;
-    if(!is_pointer_type(op0))
-      get_value_set_rec(op1, op1_set, "", op1->type, false);
+    get_value_set_rec(op0, op0_set, "", (op0)->type, false);
+    get_value_set_rec(op1, op1_set, "", (op1)->type, false);
+
+    /*
+     * The main idea here is to be able to detect if any address might be valid at all.
+     * If any side of the pointer arithmetic is not a valid-object, then we can
+     * assume that it is not the pointer operand.
+     *
+     * As of now, there is one KNOWNBUG (esbmc/github_659_2_true), which is triggered by
+     * a cast of the index as char ptr. This makes ESBMC think that both operands are
+     * valid, which is currently unsupported.
+     */
+    auto op_contains_valid_address = [this](value_sett::object_mapt o) -> bool {
+      for(const auto &it : o)
+      {
+        auto expr = object_numbering[it.first];
+        if(is_symbol2t(expr) || is_constant_string2t(expr))
```

Comment:
Are symbols and constant strings the only kind of expressions that constitute "valid addresses"? What about constant structs or unions?

Reply:
Hm, I am not sure. I will add more tests for complex data-types.

Reply:
Hm, partially. For constant array I rather meant something like [gh-695_5_true.c.txt](https://github.com/esbmc/esbmc/files/8640281/gh-695_5_true.c.txt) where the `op0_set` now also gets emptied.

Reply:
In light of the recent discussion about compound literals, I suppose my example above does not need to be supported in this PR.

However, I would still like to understand why `is_symbol2t(expr) || is_constant_string2t(expr)` is the complete condition for `expr` to constitute a "valid address". Phrased differently: what is special about `constant_string2t` that does not hold for other non-symbol expressions?